goldmane/pkg/internal/utils/files.go

```diff
@@ -15,73 +15,80 @@ package utils
 
 import (
 	"context"
+	"crypto/sha256"
+	"errors"
 	"fmt"
-	"path/filepath"
+	"os"
+	"time"
 
-	"github.com/fsnotify/fsnotify"
 	"github.com/sirupsen/logrus"
 
 	"github.com/projectcalico/calico/lib/std/chanutil"
 )
 
 // WatchFilesFn builds a closure that can be used to monitor the given files and send an update
 // to the given channel when any of the files change.
-func WatchFilesFn(updChan chan struct{}, files ...string) (func(context.Context), error) {
-	fileWatcher, err := fsnotify.NewWatcher()
-	if err != nil {
-		return nil, fmt.Errorf("error creating file watcher: %s", err)
-	}
+//
+// Note that we don't use fsnotify or a similar library here - the way that file projection in Kubernetes works relies on
+// a number of temporary files being created, copied, and removed, which can pose problems for some file watching libraries.
+// We generally don't need immediate updates on file changes, so we poll the files at a regular interval instead.
+func WatchFilesFn(updChan chan struct{}, interval time.Duration, files ...string) (func(context.Context), error) {
+	cached := map[string]string{}
 
-	watchedDirs := map[string]struct{}{}
 	for _, file := range files {
-		if file == "" {
-			continue
-		}
-
-		// There is a bug in fsnotify where some events are not triggered on files. So watch the directory -
-		// this ensures we catch all events. Longer term, moving off of fsnotify would be a good idea.
-		dir := filepath.Dir(file)
-		if _, ok := watchedDirs[dir]; ok {
-			// Already watching this directory.
-			continue
-		}
-
-		if err := fileWatcher.Add(dir); err != nil {
-			logrus.WithError(err).Warn("Error watching directory for changes")
-			continue
-		}
-		logrus.WithField("dir", dir).Debug("Watching directory for changes")
-		watchedDirs[dir] = struct{}{}
+		logrus.WithField("file", file).Debug("Watching file")
 	}
 
 	return func(ctx context.Context) {
-		for dir := range watchedDirs {
-			logrus.WithField("dir", dir).Info("Starting watch on directory")
-		}
-
-		// If we exit this function, make sure to close the file watcher and update channel.
-		defer fileWatcher.Close()
+		// If we exit this function, make sure to close the update channel.
 		defer close(updChan)
 		defer logrus.Warn("File watcher closed")
+
 		for {
-			select {
-			case <-ctx.Done():
+			if ctx.Err() != nil {
 				return
-			case event, ok := <-fileWatcher.Events:
-				if !ok {
-					logrus.Warn("File watcher events channel closed")
-					return
+			}
+
+			for _, file := range files {
```

Comment:
What would you think of calculating all the hashes first and then sending a single update if any of them has changed?